Expose trace sampling controls

douglas-reid opened the pull request:
This PR adds trace sampling controls to the public API. It is a continuation of work from PR #375.

The controls added:
- `Tracing.sampling` (equivalent to [tracing.global_enabled](https://www.envoyproxy.io/docs/envoy/latest/configuration/http_conn_man/runtime.html?highlight=global_enabled))
- `Tracing.random_sampling` (equivalent to [tracing.random_sampling](https://www.envoyproxy.io/docs/envoy/latest/configuration/http_conn_man/runtime.html?highlight=random_sampling))


Signed-off-by: Douglas Reid <douglas-reid@users.noreply.github.com>

kyessenov commented on envoy/api/v2/filter/network/http_connection_manager.proto:
```diff
@@ -80,6 +81,17 @@ message HttpConnectionManager {
+
+    // Target percentage of requests managed by this HTTP connection manager that will be randomly
+    // traced. Defaults to 100%. This variable is a direct analog for the variable of the same name
+    // in the :ref:`HTTP Connection Manager <config_http_conn_man_runtime>`.
```
I think it would be useful to add a note comparing the two fields. It's not entirely obvious to me how they differ.

mattklein123 replied:
agreed

douglas-reid replied:
I was similarly confused. I attempted a bit of clarification based on my read of the existing documentation and code. Hopefully my attempt has added clarity. Please let me know what you think.

mattklein123 commented:
FYI this PR was just closed that was working on the same thing. Please reconcile with the giant set of comments in there. Thank you. https://github.com/envoyproxy/data-plane-api/pull/375

mattklein123 commented:
Oops sorry didn't see comment on other PR. Yes if this is reconciled with that, would love to see this move forward.

douglas-reid commented:
@mattklein123 as far as I am aware, this is reconciled -- with the new bits being a move to use the newly-added `Percent` message for both fields (as suggested by @htuch).

mattklein123 commented on envoy/api/v2/filter/network/http_connection_manager.proto:
```diff
@@ -80,6 +81,17 @@ message HttpConnectionManager {
+
+    // Target percentage of requests managed by this HTTP connection manager that will be traced
+    // after all other checks have been applied (force tracing, sampling, etc.). Defaults to 100%.
+    // This variable is a direct analog for the variable named 'global_enabled' in the :ref:`HTTP
```
nit: "runtime variable"

mattklein123 commented on envoy/api/v2/filter/network/http_connection_manager.proto:
```diff
@@ -80,6 +81,17 @@ message HttpConnectionManager {
+    Percent sampling = 3;
+
+    // Target percentage of requests managed by this HTTP connection manager that will be randomly
+    // traced. Defaults to 100%. This variable is a direct analog for the variable of the same name
```
"runtime variable"

douglas-reid commented:
PTAL. I've tried to clean up the docs as suggested in comments. I also added back the `client_sampling` config bit which I inadvertently left off in the initial commit.

I've renamed `sampling` to  `overall_sampling` in an attempt to clarify (based on my perhaps erroneous understanding of how the runtime params are used).

mattklein123 commented on envoy/api/v2/filter/network/http_connection_manager.proto:
```diff
@@ -80,6 +81,30 @@ message HttpConnectionManager {
     repeated string request_headers_for_tags = 2;
+
+    // Target percentage of requests managed by this HTTP connection manager that will be force
+    // traced if the *x-client-trace-id* header is set. This field is a direct analog for the
```
please sphinx ref here: https://www.envoyproxy.io/docs/envoy/latest/configuration/http_conn_man/headers#x-client-trace-id

douglas-reid replied:
Done.

wora commented:
Thanks for the detailed documentation.

/lgtm

douglas-reid commented:
Thanks for the quick reviews!

I've added the ref for x-client-trace-id as requested. PTAL.
